### ISSUE TITLE
build: document is not defined

### ISSUE THREAD

**Issue by Simon-He95** [antfu/vite-ssg#254]
![image](https://user-images.githubusercontent.com/57086651/171075977-8bd01759-c71b-468f-9231-6ed80f7bb650.png)
Under such scenario,I introduced an npm package component that relies on document

This error occurs that vite-ssg build

**Comment by blwtxc** [comment]
Stumbled upon that problem several times now. Before I was able to wrap it and create a mock *window/document* to make it build. Now I added an npm package that wants *document* (quill.js in my case) and I don't know what to do now. Project runs smooth in *dev* but wont *build*. Is there perhaps an option I overlooked that injects a mock document to make modules build?
Cheers for any help from the community :)

**Comment by Simon-He95** [comment]
![image](https://user-images.githubusercontent.com/57086651/171196922-71d4959c-c540-4f4c-9bb6-2c403287bb8e.png)
I want to create style element append head, In this case will build wrong.
I can only use it by using onBeforeMount to wrap it in a component, it will be ok, but It wasn't what I expected.The component will be used many times,so i need to add additional conditions to control unexpected execution.

**Comment by blwtxc** [comment]
@antfu any idea? did I miss an option?

**Comment by Simon-He95** [comment]
I try to wrap the problematic code portion(s) into try .. catch to handle, the issue was resolved.
Thank you for your help, but I'm more looking forward to being able to create a mock window/document to make it build.

**Comment by blwtxc** [comment]
... I just had another look at the issues-page and this problem is prevalent when using vite-ssg, but it seems to be also solved (?) ... I have to take a look at their suggestions and see if they are feasable for my situation ... did you have a look there and tried some of their suggestions?

**Comment by blwtxc** [comment]
... and ... yes, it would of course be awesome, if that would be something taken care of on vite-ssg's side ;)

**Comment by userquin** [comment]
@blwtxc I sent a PR a few days ago to allow use `jsdom`, the problem is that the `mock: true` option enables it usage but `jsdom` is registered after building the SSR main and may fail

**Comment by Simon-He95** [comment]
Yeah,most of the problems are in build process, when build we might be able to wrap window/document into try...catch or mock window/document, If there are errors, let the errors be thrown at runtime

**Comment by userquin** [comment]
@Simon-He95 the problem here is the executon of `main.[cm]js` not its generation: if you're using some lib referencing window/document it will fail while importing that library when we run it, just check the `vite-ssg-temp` folder and the main entry generated.

This line https://github.com/antfu/vite-ssg/blob/main/src/node/build.ts#L107 will load previous generated module, but the `jsdom` registration is done a few lines below it, we need to register jsdom (when using `mock: true`) at least before line 107, my PR registers the jsdom before its generation

**Comment by userquin** [comment]
you can also excluded that libraries using `optimizedDeps.exclude` and try build: for example, while testing vite 3 with vitesse template I need to exclude `workbox-window`

![imagen](https://user-images.githubusercontent.com/6311119/171389527-f3c76452-7c4a-4c40-9427-16d2185087da.png)

**Comment by blwtxc** [comment]
@userquin thanks a lot for chiming in and your reply!! :D

**Comment by blwtxc** [comment]
@userquin Thanks! Already achieved it with `happy-dom` in the `build.[cm]js` module. I guess I will try it with your changes the next time I have to touch the modules or I realize that something's not working after all :))

**Comment by blwtxc** [comment]
@userquin Ok, thanks. If you want, I can let u know if I run into problems with `happy-dom` ... kinda case study ;)

**Comment by userquin** [comment]
@blwtxc  what do you mean, adding also happy-dom to use it on SSG instead jsdom? If so, maybe you can change `mock` to be `mock?: 'happy-dom' | 'jsdom' | boolean` , when configuring `mock: true` fallback to 'jsdom' and so no breaking change, but you must include some common interface to do all current ssg work using happy-dom

I'll move this issue to a discussion, my PR should fix this

**Comment by blwtxc** [comment]
@userquin I think that is a brilliant idea!
What do you mean by *include some common interface*?
(Sorry, maybe that's a stupid question...)

**Comment by userquin** [comment]
@blwtxc we should use/replace from this line https://github.com/antfu/vite-ssg/blob/main/src/node/build.ts#L167 to this another https://github.com/antfu/vite-ssg/blob/main/src/node/build.ts#L175 (both included, I'm assuming happy-dom can serialize the dom to html)

**Comment by blwtxc** [comment]
@userquin I see now where I made a mistake.
Using `GlobalRegistrator.register()` was only an easy fix that allowed modules in my app to run without throwing errors about missing DOM elements and functions. Based on that I assumed that `happy-dom` replaced `jsdom`, but it did not. The heavy lifting is still done by `jsdom` as it seems.
I can't find a serialize method, that does the same as the one provided by `jsdom`. I guess `happy-dom` could be used as a replacement, but I don't see a solution that wouldn't include rewriting the whole build module... and I don't think that is something worth spending timee on...

**Comment by userquin** [comment]
@blwtxc continues here https://github.com/antfu/vite-ssg/discussions/256
